### PR TITLE
Configurably optional warnings

### PR DESCRIPTION
**Pull request by moskyb**
### Description

there's a subset of warnings in the agent that are essentially us trying to clippy the users, saying "Hmm, something about this setup looks a little bit weird, but we'll proceed anyway" - notably a message about when [submodules are disabled, but the repo the agent is cloning containing submodules](https://github.com/buildkite/agent/issues/1773).

he somewhat frequently get feedback from users that when you know that you're doing the right thing - for example, if you know you're not going to need the submodules that aren't being cloned - these warnings are annoying, and pop open a large log group in the job log that isn't useful.

however, we're usually a bit reluctant to add flags to disable individual warnings - for example, a hypothetical `--disable-warning-about-submodules`, because the agent arguably has too many flags already.

this PR is a bit of a middle ground - certain warnings will become optional by using the `--no-warn-for` flag. Each optional warning is given an ID, and when not disabled, will have  "You can disable this warning by passing the `--no-warn-for <some warning name>`" appended to the end.

If users want to disable that warning - for example, the submodule warning mentioned above, `submodule-conflict` - they can pass `--no-warn-for submodule-conflict` to their `buildkite-agent start` command.

Thanks so much to @amu-g for pairing with me on this!

### Testing
- [x] Tests have run locally (with `go test ./...`). Buildkite employees may check this if the pipeline has run automatically.
- [x] Code is formatted (with `go fmt ./...`)

### CODE REVIEW

File: internal/job/shell/logger.go
Comment (moskyb) [inline internal/job/shell/logger.go:107]: one weakness of this approach is that the values of optional warnings aren't easily discoverable, and could potentially intersect.

one idea we had for this was to make a warning id type and use that instead, a la
```Go
type optionalWarningID int
const (
  JobAPIDisabled optionalWarningID = iota
  SubmoduleConflict
  // ...etc
)
```

and then make this function take an `optionalWarningID` for the `id` param rather than a string. the reason that we ended up going with this approach is that _which_ warnings are ignored are more a concern of the job executor than they are of this module, but that decision isn't without its drawbacks.

WDYT?

File: clicommand/agent_start.go
Comment (dabarrell) [inline clicommand/agent_start.go:658]: I can't review the go, but `no-warn-for` feels weird - grammatically incorrect etc. Seems we're okay with slightly longer flags if it means they read more easily. How about `disable-warnings-for`? to match the language you're using in the usage string, and the code (e.g. `DisabledWarningIDs`)

File: internal/job/shell/logger.go
Comment (dabarrell) [inline internal/job/shell/logger.go:101]: This is what I came to check - whether or not you downgrade to `info` level log line, or drop it entirely. Perhaps a `debug` line is a good middle ground? discoverable if debug is enabled, otherwise not?

File: internal/job/shell/logger.go
Comment (DrJosh9000) [reply internal/job/shell/logger.go:107]: How likely is it that someone adding a new optional warning is going to accidentally re-use an existing ID?

File: clicommand/agent_start.go
Comment (moskyb) [reply clicommand/agent_start.go:658]: yeah, that feels way better

File: internal/job/shell/logger.go
Comment (moskyb) [reply internal/job/shell/logger.go:101]: ah, logs don't quite work that way (with levels etc) in the shell, they're either there or they're not - i think we should probably downgrade it to just a call to `wl.Printf` though.

Comment (dabarrell) [reply internal/job/shell/logger.go:101]: Ah, okay. Yeah, I think printing something is better than nothing, and if that itself proves to be a pain point for people, we can evaluate dropping them entirely then

File: internal/job/shell/logger.go
Comment (moskyb) [reply internal/job/shell/logger.go:107]: relatively unlikely, i think, and i think the impact would be relatively low? ie they'd hide more warnings than they were intending to

File: clicommand/bootstrap.go
Comment (bpoland) [inline clicommand/bootstrap.go:369]: Should the env var be renamed too? It's certainly longer but hat seems less confusing as a user if the CLI arg and env var match
```suggestion
			EnvVar: "BUILDKITE_AGENT_DISABLE_WARNINGS_FOR",
```

Comment (moskyb) [reply clicommand/bootstrap.go:369]: yes def! good point, forgot that in my find-and-replace spree 😅

**Review by bpoland — APPROVED** [review]
This looks great to me! Do the docs on your site get generated automatically after this is merged?

**Comment by HugeIRL** [comment]
> This looks great to me! Do the docs on your site get generated automatically after this is merged?

Hey @bpoland,

The docs may pick up some of the changes, however us from Support will make sure that if anything is missing/not clear, we can jump on that to fix it.

**Review by DrJosh9000 — APPROVED** [review]
Inventory Get: Green Tick of Approval ✅

**Pull request merged**
